Svc/FileDownlink/FileDownlink.cpp

```diff
@@ -49,160 +58,268 @@ namespace Svc {
+    // the file to be downlinked anyway
+    if (startOffset >= this->file.size) {
+        this->enterIdle();
+        this->log_WARNING_HI_DownlinkPartialFail(this->file.sourceName, this->file.destName, startOffset, this->file.size);
+        this->cmdResponse_out(opCode, cmdSeq, Fw::COMMAND_VALIDATION_ERROR);
+        return;
+    } else if (startOffset + length > this->file.size) {
+        this->log_WARNING_LO_DownlinkPartialWarning(startOffset, length, this->file.size, this->file.sourceName, this->file.destName);
+        length = this->file.size - startOffset;
     }
 
-    this->mode.set(Mode::IDLE);
-    
-    this->cmdResponse_out(opCode, cmdSeq, Fw::COMMAND_OK);
+    // Store the current command info for completion later
+    this->curOpCode = opCode;
+    this->curCmdSeq = cmdSeq;
+    this->getBuffer(this->buffer, FILE_PACKET);
 
+    // Send file and switch to WAIT mode
+    this->sendStartPacket();
+    this->mode.set(Mode::WAIT);
+    this->sequenceIndex = 1;
+    this->curTimer = 0;
+    this->byteOffset = startOffset;
+    this->lastCompletedType = Fw::FilePacket::T_START;
+
+    // zero length means read until end of file
+    if (length > 0) {
+        this->log_ACTIVITY_HI_SendStarted(length, this->file.sourceName, this->file.destName);
+        this->endOffset = startOffset + length;
+    }
+    else {
+        this->log_ACTIVITY_HI_SendStarted(this->file.size - startOffset, this->file.sourceName, this->file.destName);
+        this->endOffset = this->file.size;
+    }
   }
 
   void FileDownlink ::
-    FileDownlink_Cancel_cmdHandler(
+    Cancel_cmdHandler(
         const FwOpcodeType opCode,
         const U32 cmdSeq
     )
   {
-    if (this->mode.get() == Mode::DOWNLINK)
-      this->mode.set(Mode::CANCEL);
-    this->cmdResponse_out(opCode, cmdSeq, Fw::COMMAND_OK);
+      //Must be able to cancel in both downlink and waiting states
+      if (this->mode.get() == Mode::DOWNLINK || this->mode.get() == Mode::WAIT) {
+          this->mode.set(Mode::CANCEL);
+      }
+      this->cmdResponse_out(opCode, cmdSeq, Fw::COMMAND_OK);
   }
 
   // ----------------------------------------------------------------------
   // Private helper methods 
   // ----------------------------------------------------------------------
 
   Os::File::Status FileDownlink ::
-    sendDataPacket(const U32 byteOffset)
+    sendDataPacket(U32 &byteOffset)
   {
-
-    const U32 fileSize = this->file.size;
-    FW_ASSERT(byteOffset < fileSize, byteOffset);
-    const U16 maxDataSize = static_cast<U16>(this->downlinkPacketSize);
-    const U16 dataSize = static_cast<U16>((byteOffset + maxDataSize > fileSize) ?
-      fileSize - byteOffset : maxDataSize);
+    FW_ASSERT(byteOffset < this->endOffset);
+    const U32 maxDataSize = FILEDOWNLINK_INTERNAL_BUFFER_SIZE - Fw::FilePacket::DataPacket::HEADERSIZE;
+    const U32 dataSize = (byteOffset + maxDataSize > this->endOffset) ?
```

Comment:
Nitpick, but a ternary operator wrapped across multiple lines like this is difficult to read. Maybe break into an if statement?